apps/google-analytics-4/lambda/src/lib/verify-signed-request-middleware.ts

```diff
@@ -0,0 +1,87 @@
+import * as NodeAppsToolkit from '@contentful/node-apps-toolkit';
+import {
+  CanonicalRequest,
+  ContentfulContextHeader,
+  ContentfulHeader,
+} from '@contentful/node-apps-toolkit/lib/requests/typings';
+import express from 'express';
+import { InvalidSignature } from '../lib/errors/invalid-signature';
+import { UnableToVerifyRequest } from '../lib/errors/unable-to-verify-request';
+import { IncomingHttpHeaders } from 'http';
+
+export const verifySignedRequestMiddleware = (
+  req: express.Request,
+  _res: express.Response,
+  next: express.NextFunction
+) => {
+  const signingSecret = (process.env.SIGNING_SECRET || '').trim();
+  const canonicalReq = makeCanonicalReq(req);
+  let isValidReq = false;
+
+  try {
+    isValidReq = NodeAppsToolkit.verifyRequest(signingSecret, canonicalReq, 60); // 60 second TTL
+  } catch (e) {
+    console.error(e);
+    throw new UnableToVerifyRequest('Unable to verify request', {
+      cause: e,
+    });
+  }
+
+  if (!isValidReq) {
+    throw new InvalidSignature(
+      'Request does not have a valid request signature. ' +
+        'See: https://www.contentful.com/developers/docs/extensibility/app-framework/request-verification/'
+    );
+  }
+
+  next();
+};
+
+const makeCanonicalReq = (req: express.Request) => {
+  const requiredHeaders = requestSigningHeaders(req.headers);
+
+  return <CanonicalRequest>{
+    method: req.method,
+    path: `/dev${req.originalUrl}`, // TODO: wtf express? how do i get the real request path?!
+    headers: requiredHeaders,
+  };
+};
+
+const contentfulSigningHeaderKeys = [
+  ...Object.values(ContentfulHeader),
+  ...Object.values(ContentfulContextHeader),
+];
+
+interface ContentfulSignedHeaders {
+  'x-contentful-timestamp': string;
+  'x-contentful-signed-headers': string;
+  'x-contentful-signature': string;
+  'x-contentful-user-id': string;
+  'x-contentful-space-id': string;
+  'x-contentful-environment-id': string;
+  'x-contentful-app-id': string;
+  [key: string]: string;
+}
+
+// TODO: can we refactor to sth like this?
```

Comment:
I'd ditto capturing the TODO as a task or maybe leaving it out otherwise. (No strong feelings either way)

Reply:
oh i left this as a question i wanted to ask @pkeavenycontentful for input/pairing on! I couldn't get this working for the life of me 😅

Reply:
(removed for now and will pair with Patrick to see what he thinks!)